staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -29,6 +29,44 @@ type Token interface {
 	AuthenticateToken(ctx context.Context, token string) (*Response, bool, error)
 }
 
+func WrapInInstrumentedRequest(req Request, authenticatorName string) *InstrumentedRequest {
+	return &InstrumentedRequest{Name: authenticatorName, delegate: req}
+}
+
+// InstrumentedRequest wraps Request and allows instrumentation to identify authenticators via the Name field.
+type InstrumentedRequest struct {
+	// Name identifies the delegate authenticator.
+	Name     string
+	delegate Request
+}
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *InstrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	if err != nil {
+		err = InstrumentedError{Name: i.Name, delegate: err}
+	}
+
+	if resp != nil {
+		resp.AuthenticatorName = i.Name
+	}
+
+	return resp, ok, err
+}
+
+// InstrumentedError implementation of the Error interface, and is intended for use by the
+// InstrumentedRequest.
+type InstrumentedError struct {
+	// Name identifies the delegate authenticator that produced the error.
+	Name     string
+	delegate error
+}
+
+// Error implementation of the Error interface.
+func (e InstrumentedError) Error() string {
```

Comment:
nit: if you embed the delegate, you don't need this function.

Reply:
Done.